### ISSUE TITLE
Missing variables from variable map in second order coupled PDE - but only if fully coupled (?)

### ISSUE THREAD

**Issue by ppinney** [SciML/ModelingToolkit.jl#2924]
Hello, I am trying to solve a large set of coupled PDEs using MethodOfLines.jl and ModelingToolkit.jl, and am running into discretization issues. A simplified version of the problem (with the same error) is as follows:

```
using MethodOfLines, ModelingToolkit, OrdinaryDiffEq, DomainSets

@parameters x t
@variables ϕ1(..) ϕ2(..)
Dt = Differential(t)
Dx = Differential(x)
Dtt = Differential(t)^2
Dxx = Differential(x)^2

γ = 0.01
α = 0.05
S = -0.499997
l = 5

bc = 0

x_min = t_min = 0
x_max = l
t_max = 20


J(ϕ) = Dtt(ϕ) + α*Dt(ϕ) - γ

eq = [Dxx(ϕ1(x,t)) ~ J(ϕ1(x,t)) + S*J(ϕ2(x,t)),
      Dxx(ϕ2(x,t)) ~  S*J(ϕ1(x,t)) + J(ϕ2(x,t))]

domains = [x ∈ Interval(x_min, x_max),
      t ∈ Interval(t_min, t_max)]


bcs = [Dx(ϕ1(x_min,t)) ~ bc,
      Dx(ϕ1(x_max,t)) ~ bc,
      ϕ1(x,t_min) ~ bc,
      Dt(ϕ1(x,t_min)) ~ bc,
      Dx(ϕ2(x_min,t)) ~ bc,
      Dx(ϕ2(x_max,t)) ~ bc,
      ϕ2(x,t_min) ~ bc,
      Dt(ϕ2(x,t_min)) ~ bc]

@named pdesys = PDESystem(eq,bcs,domains,[x,t],[ϕ1(x,t),ϕ2(x,t)])

Nd = 8
order = 2
discretization = MOLFiniteDifference([x=>Nd],t,approx_order=order)

println("Discretization:")
@time prob = discretize(pdesys,discretization)
```

When I run this, I receive the following error message:
```
ArgumentError: SymbolicUtils.BasicSymbolic{Real}[var"ϕ1(t)[2]ˍtt", var"ϕ1(t)[3]ˍtt", var"ϕ1(t)[4]ˍtt", var"ϕ1(t)[5]ˍtt", var"ϕ1(t)[6]ˍtt", var"ϕ2(t)[7]ˍtt"] are missing from the variable map.

Stacktrace:
  [1] throw_missingvars(vars::Vector{SymbolicUtils.BasicSymbolic{Real}})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/variables.jl:122
  [2] _varmap_to_vars(varmap::Dict{Any, Any}, varlist::Vector{SymbolicUtils.BasicSymbolic{Real}}; defaults::Dict{Any, Any}, check::Bool, toterm::typeof(ModelingToolkit.default_toterm))
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/variables.jl:116
  [3] varmap_to_vars(varmap::Vector{Pair}, varlist::Vector{SymbolicUtils.BasicSymbolic{Real}}; defaults::Dict{Any, Any}, check::Bool, toterm::Function, promotetoconcrete::Nothing, tofloat::Bool, use_union::Bool)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/variables.jl:85
  [4] get_u0_p(sys::ODESystem, u0map::Vector{Pair}, parammap::SciMLBase.NullParameters; use_union::Bool, tofloat::Bool, symbolic_u0::Bool)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:787
  [5] get_u0_p
    @ ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:769 [inlined]
  [6] process_DEProblem(constructor::Type, sys::ODESystem, u0map::Vector{Pair}, parammap::SciMLBase.NullParameters; implicit_dae::Bool, du0map::Nothing, version::Nothing, tgrad::Bool, jac::Bool, checkbounds::Bool, sparse::Bool, simplify::Bool, linenumbers::Bool, parallel::Symbolics.SerialForm, eval_expression::Bool, use_union::Bool, tofloat::Bool, symbolic_u0::Bool, u0_constructor::typeof(identity), kwargs::@Kwargs{t::Int64, has_difference::Bool, check_length::Bool})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:812
  [7] (ODEProblem{true, SciMLBase.AutoSpecialize})(sys::ODESystem, u0map::Vector{Pair}, tspan::Tuple{Int64, Int64}, parammap::SciMLBase.NullParameters; callback::Nothing, check_length::Bool, kwargs::@Kwargs{})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:937
  [8] ODEProblem
    @ ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:930 [inlined]
  [9] (ODEProblem{true, SciMLBase.AutoSpecialize})(sys::ODESystem, u0map::Vector{Pair}, tspan::Tuple{Int64, Int64})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:930
 [10] (ODEProblem{true})(::ODESystem, ::Vector{Pair}, ::Vararg{Any}; kwargs::@Kwargs{})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:917
 [11] (ODEProblem{true})(::ODESystem, ::Vector{Pair}, ::Vararg{Any})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/Nsxrf/src/systems/diffeqs/abstractodesystem.jl:916
...
    @ PDEBase ~/.julia/packages/PDEBase/Aqj4G/src/discretization_state.jl:74
 [15] discretize(pdesys::PDESystem, discretization::MOLFiniteDifference{MethodOfLines.CenterAlignedGrid, MethodOfLines.ScalarizedDiscretization})
    @ PDEBase ~/.julia/packages/PDEBase/Aqj4G/src/discretization_state.jl:55
```

Based on other posts I've found online, it seems like this should have to do with missing initial conditions, but I've tried adding more and it didn't help. Interestingly, if I even partially decouple the equations, e.g.:
```
eq = [Dxx(ϕ1(x,t)) ~ J(ϕ1(x,t)),
      Dxx(ϕ2(x,t)) ~  S*J(ϕ1(x,t)) + J(ϕ2(x,t))]
```
or
```
eq = [Dxx(ϕ1(x,t)) ~ J(ϕ1(x,t)) + S*J(ϕ2(x,t)),
      Dxx(ϕ2(x,t)) ~  J(ϕ2(x,t))]
```
the discretization proceeds, and the system can be solved!

If anyone can help identify what the issue is, any help would be greatly appreciated.

**Comment by ppinney** [comment]
I realized I hadn't updated packages in a while, and I after updating the error message changed to to be more clear:
```
Initial condition underdefined. Some are missing from the variable map.
Please provide a default (`u0`), initialization equation, or guess
for the following variables:

Any[var"ϕ2(t)[2]ˍtt", var"ϕ1(t)[3]ˍtt", var"ϕ1(t)[4]ˍtt", var"ϕ2(t)[5]ˍtt", var"ϕ1(t)[6]ˍtt", var"ϕ1(t)[7]ˍtt"]


Stacktrace:
  [1] _varmap_to_vars(varmap::Dict{Any, Any}, varlist::Vector{SymbolicUtils.BasicSymbolic{Real}}; defaults::Dict{Any, Any}, check::Bool, toterm::Function, initialization_phase::Bool)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/variables.jl:212
  [2] varmap_to_vars(varmap::Vector{Pair}, varlist::Vector{SymbolicUtils.BasicSymbolic{Real}}; defaults::Dict{Any, Any}, check::Bool, toterm::Function, promotetoconcrete::Nothing, tofloat::Bool, use_union::Bool)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/variables.jl:162
  [3] get_u0(sys::ODESystem, u0map::Vector{Pair}, parammap::SciMLBase.NullParameters; symbolic_u0::Bool, toterm::Function)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:0
  [4] process_DEProblem(constructor::Type, sys::ODESystem, u0map::Vector{Pair}, parammap::SciMLBase.NullParameters; implicit_dae::Bool, du0map::Nothing, version::Nothing, tgrad::Bool, jac::Bool, checkbounds::Bool, sparse::Bool, simplify::Bool, linenumbers::Bool, parallel::Symbolics.SerialForm, eval_expression::Bool, eval_module::Module, use_union::Bool, tofloat::Bool, symbolic_u0::Bool, u0_constructor::typeof(identity), guesses::Dict{Any, Any}, t::Int64, warn_initialize_determined::Bool, build_initializeprob::Bool, initialization_eqs::Vector{Any}, fully_determined::Bool, kwargs::@Kwargs{check_length::Bool})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:857
  [5] process_DEProblem
    @ ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:751 [inlined]
  [6] (ODEProblem{true, SciMLBase.AutoSpecialize})(sys::ODESystem, u0map::Vector{Pair}, tspan::Tuple{Int64, Int64}, parammap::SciMLBase.NullParameters; callback::Nothing, check_length::Bool, warn_initialize_determined::Bool, eval_expression::Bool, eval_module::Module, kwargs::@Kwargs{build_initializeprob::Bool})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:1009
  [7] ODEProblem
    @ ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:997 [inlined]
  [8] (ODEProblem{true})(::ODESystem, ::Vector{Pair}, ::Vararg{Any}; kwargs::@Kwargs{build_initializeprob::Bool})
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/yHDgG/src/systems/diffeqs/abstractodesystem.jl:984
  [9] ODEProblem(::ODESystem, ::Vector{Pair}, ::Vararg{Any}; kwargs::@Kwargs{build_initializeprob::Bool})
...
    @ PDEBase ~/.julia/packages/PDEBase/nzap9/src/discretization_state.jl:74
 [11] discretize(pdesys::PDESystem, discretization::MOLFiniteDifference{MethodOfLines.CenterAlignedGrid, MethodOfLines.ScalarizedDiscretization})
    @ PDEBase ~/.julia/packages/PDEBase/nzap9/src/discretization_state.jl:55
```
However, I still can't solve this issue. I'm not sure where I should be adding defaults, initialization equations, or guesses. I've tried adding BCs that depend on other equations, I've tried adding defaults to the PDESystem like in this line:
```
pdesys = PDESystem(eq,bcs,domains,[x,t],[ϕ1(x,t),ϕ2(x,t)],defaults = Dict(Dtt(ϕ1(x,t)) => γ,Dtt(ϕ2(x,t)) => γ))
```
but nothing I can think of seems to work.

**Comment by ppinney** [comment]
Okay, I ended up figuring out a workaround; not one that I'd prefer, as it seems to impact performance, but I can now discretize and find a solution.

After following the call stack and tooling with `structural_simplify`, it seemed that the error had specifically to do with time, and not space. So, I tried treating time and space equivalently, and it seems to have worked! Specifically, before `discretize()` is called, I altered the `MOLFiniteDifference` call to be:
```
discretization = MOLFiniteDifference([x=>Nd,t=>Nd],approx_order=order)
```
and it seems to work! Now I'm just working on getting it to solve in a reasonable time.
